File: leetcode/79.word-search.py
# @before-stub-for-debug-begin
from python3problem79 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=79 lang=python3
#
# [79] Word Search
#

# @lc code=start
class Solution:
    def exist(self, board, word):
        rows, cols = len(board), len(board[0])
        path = set()

        def dfs(r, c, i):
            if i == len(word):
                return True

            if (
                r < 0
                or c < 0
                or r >= rows
                or c >= cols
                or word[i] != board[r][c]
                or (r, c) in path
            ):
                return False

            path.add((r, c))

            res = (
                dfs(r + 1, c, i + 1)
                or dfs(r, c + 1, i + 1)
                or dfs(r - 1, c, i + 1)
                or dfs(r, c - 1, i + 1)
            )
            path.remove((r, c))
            return res

        for r in range(rows):
            for c in range(cols):
                if dfs(r, c, 0):
                    return True
        # O(n*m*dfs times) = O(n*m*4^len(word))
        # dfs O(len(word)
        # 4^len(word)


# @lc code=end

File: leetcode/79.word-search.java

/*
 * @lc app=leetcode id=79 lang=java
 *
 * [79] Word Search
 */

// @lc code=start
class Solution {
    public boolean exist(char[][] board, String word) {
        int r = board.length;
        int c = board[0].length;

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static final int[][] DIRS = new int[][] {
            { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };

    private boolean dfs(char[][] board, String word, int r, int c, int i) {
        if (i == word.length()) {
            return true;
        }
        if (r < 0 || c < 0 || r >= board.length
                || c >= board[0].length
                || word.charAt(i) != board[r][c]
                || board[r][c] == '#') {
            return false;
        }

        char temp = board[r][c];
        board[r][c] = '#';
        boolean res = false;
        for (var dir : DIRS) {
            if (dfs(board, word, r + dir[0], c + dir[1], i + 1)) {
                res = true;
            }
        }
        board[r][c] = temp;
        return res;
    }
}
// @lc code=end
